src/main/java/org/datadog/jenkins/plugins/datadog/traces/write/TraceWriter.java

```diff
@@ -0,0 +1,118 @@
+
+public class TraceWriter {
+
+    private static final Logger logger = Logger.getLogger(TraceWriter.class.getName());
+
+    private static final String QUEUE_CAPACITY_ENV_VAR = "DD_JENKINS_TRACES_QUEUE_CAPACITY";
+    private static final String SUBMIT_TIMEOUT_ENV_VAR = "DD_JENKINS_TRACES_SUBMIT_TIMEOUT_SECONDS";
+    private static final String STOP_TIMEOUT_ENV_VAR = "DD_JENKINS_TRACES_STOP_TIMEOUT_SECONDS";
+    private static final String POLLING_INTERVAL_ENV_VAR = "DD_JENKINS_TRACES_POLLING_INTERVAL_SECONDS";
+    private static final String BATCH_SIZE_LIMIT_ENV_VAR = "DD_JENKINS_TRACES_BATCH_SIZE_LIMIT";
+    private static final int DEFAULT_QUEUE_CAPACITY = 10_000;
+    private static final int DEFAULT_SUBMIT_TIMEOUT_SECONDS = 30;
+    private static final int DEFAULT_STOP_TIMEOUT_SECONDS = 15;
+    private static final int DEFAULT_POLLING_INTERVAL_SECONDS = 10;
+    private static final int DEFAULT_BATCH_SIZE_LIMIT = 100;
+
+    private final TraceWriteStrategy traceWriteStrategy;
+    private final BlockingQueue<JSONObject> queue;
+    private final Thread poller;
+
+    public TraceWriter(DatadogClient datadogClient) {
+        this.traceWriteStrategy = datadogClient.createTraceWriteStrategy();
+
+        this.queue = new ArrayBlockingQueue<>(getEnv(QUEUE_CAPACITY_ENV_VAR, DEFAULT_QUEUE_CAPACITY));
+
+        this.poller = new Thread(new QueuePoller(traceWriteStrategy, queue), "DD-Trace-Writer");
+        this.poller.setDaemon(true);
+        this.poller.start();
+    }
+
+    public void submitBuild(final BuildData buildData, final Run<?,?> run) throws InterruptedException, TimeoutException {
+        JSONObject buildJson = traceWriteStrategy.serialize(buildData, run);
+        submit(buildJson);
+    }
+
+    public void submitPipeline(FlowNode flowNode, Run<?, ?> run) throws InterruptedException, TimeoutException {
+        Collection<JSONObject> nodeJsons = traceWriteStrategy.serialize(flowNode, run);
+        for (JSONObject nodeJson : nodeJsons) {
+            submit(nodeJson);
+        }
+    }
+
+    private void submit(JSONObject json) throws InterruptedException, TimeoutException {
+        if (!queue.offer(json, getEnv(SUBMIT_TIMEOUT_ENV_VAR, DEFAULT_SUBMIT_TIMEOUT_SECONDS), TimeUnit.SECONDS)) {
+            throw new TimeoutException("Timed out while submitting span");
+        }
+    }
+
+    public void stop() {
+        poller.interrupt();
+    }
+
+    private static final class QueuePoller implements Runnable {
+        private final TraceWriteStrategy traceWriteStrategy;
+        private final BlockingQueue<JSONObject> queue;
+
+        public QueuePoller(TraceWriteStrategy traceWriteStrategy, BlockingQueue<JSONObject> queue) {
+            this.traceWriteStrategy = traceWriteStrategy;
+            this.queue = queue;
+        }
+
+        @Override
+        public void run() {
+            long shutdownAt = Long.MAX_VALUE;
+            while (System.currentTimeMillis() < shutdownAt) {
```

Comment:
What happens if the thread must stop, but there are spans in the queue pending to be sent? Should we take into account the pending payloads in the queue as well?

Reply:
When the interrupt is received, `shutdownAt` is set to 15 seconds in the future. So we continue polling/flushing for the duration of this "shutdown period" to make sure pending spans get flushed. It is limited, however, to not delay the shutdown too much (e.g. if the backend is slow to respond, we want to wait for a reasonable amount of time and eventually stop, rather than delaying the shutdown for an arbitrarily long period).